units/src/amount.rs

```diff
@@ -145,14 +145,55 @@ impl FromStr for Denomination {
     }
 }
 
+
+/// An error during amount parsing amount with denomination.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum ParseError {
+    /// Invalid amount.
+    Amount(ParseAmountError),
+
+    /// Invalid denomination.
+    Denomination(ParseDenominationError),
+}
+
+impl From<ParseAmountError> for ParseError {
+    fn from(e: ParseAmountError) -> Self { Self::Amount(e) }
+}
+
+impl From<ParseDenominationError> for ParseError {
+    fn from(e: ParseDenominationError) -> Self { Self::Denomination(e) }
+}
+
+impl From<OutOfRangeError> for ParseError {
+    fn from(e: OutOfRangeError) -> Self { Self::Amount(e.into()) }
+}
+
+impl fmt::Display for ParseError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            ParseError::Amount(error) => write_err!(f, "invalid amount"; error),
+            ParseError::Denomination(error) => write_err!(f, "invalid denomination"; error),
+        }
```

Comment:
Any reason why this uses different style to other code we have e.g.,
```rust
use ParseError::*;

match *self {
            Amount(error) => write_err!(f, "invalid amount"; error),
            Denomination(error) => write_err!(f, "invalid denomination"; error),
        }
```

In hindsighte I kind of wish we used `use Self::*` instead of naming the type to make maintenance slightly cheaper.

Reply:
Note please this is not a my style vs your style thing, I think the style we use is clearer. (I intentionally did not change `error` to `e`.)

Reply:
Various reasons:

* I actually wrote this a long time ago
* I didn't notice we've adopted this as official style
* I don't mind typing the enum name so I wasn't trying to get rid of it

I think Swift got this one better having `.Foo` meaning "variant Foo of self", I wish this was available in Rust somehow!

Reply:
> In hindsight I kind of wish we used `use Self::*`

I checked and this doesn't work, that is a surprise to me.

Reply:
Yeah, I hit it recently in another project. Was surprised as well.